### PR TITLE
Migrates prepay facet from space factory to space

### PR DESCRIPTION
**Pull request by giuseppecrj**
Context:
* The current prepay facet implementation doesn't take into account the reduction of prepaid memberships once minted, we end up with inaccurate accounting mechanism. 
* An example is a town owner prepays for 2 memberships and potential members mint those memberships, I still hold the accounting of 2 memberships in our prepay storage, which can lead to exploits.
* It makes more sense to move this logic to the space itself since the proper reduction of prepaid memberships can be done within the `_issueToken` functionality of our membership minting process

Proposals:
* https://testnets.llama.xyz/orgs/river/base-sepolia/actions/126
* https://testnets.llama.xyz/orgs/river/base-sepolia/actions/125

### CODE REVIEW

File: contracts/src/spaces/facets/prepay/PrepayFacet.sol [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  function prepayMembership(
+    uint256 supply
+  ) external payable nonReentrant onlyOwner {
+    if (supply == 0) revert Prepay__InvalidSupplyAmount();
+
+    MembershipStorage.Layout storage ds = MembershipStorage.layout();
+    IPlatformRequirements platform = IPlatformRequirements(ds.spaceFactory);
+
+    uint256 cost = supply * platform.getMembershipFee();
```
Comment (pfives) [inline contracts/src/spaces/facets/prepay/PrepayFacet.sol:31]: why not using calculateMembershipPrepayFee function defined below? if not why keep that?

**Review by pfives — APPROVED** [review]

File: contracts/src/spaces/facets/prepay/PrepayFacet.sol (hunk 1 above)
Comment (giuseppecrj) [reply contracts/src/spaces/facets/prepay/PrepayFacet.sol:31]: good catch! will update.

File: contracts/src/spaces/facets/membership/MembershipFacet.sol [hunk 2]
```diff
@@ -147,8 +150,22 @@ contract MembershipFacet is
       (address, address)
     );
 
-    // allocate protocol and membership fees
-    uint256 membershipPrice = _getMembershipPrice(_totalSupply());
+    uint256 totalSupply = _totalSupply();
+    uint256 membershipPrice;
+
+    uint256 freeAllocation = _getMembershipFreeAllocation();
+    uint256 prepaidSupply = _getPrepaidSupply();
```
Comment (salzbrenner) [inline contracts/src/spaces/facets/membership/MembershipFacet.sol:157]: can you remind me what is the difference between freeAllocation and prepaidSupply?

File: contracts/src/spaces/facets/prepay/PrepayBase.sol [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  function _addPrepay(uint256 supply) internal {
+    PrepayStorage.Layout storage ds = PrepayStorage.layout();
+    ds.supply += supply;
+    emit Prepay__Prepaid(supply);
+  }
+
+  function _reducePrepay(uint256 supply) internal {
+    PrepayStorage.Layout storage ds = PrepayStorage.layout();
+    ds.supply -= supply;
+  }
```
Comment (salzbrenner) [inline contracts/src/spaces/facets/prepay/PrepayBase.sol:22]: is there any theoretical situation where supply could become < 0?

File: contracts/src/spaces/facets/membership/MembershipFacet.sol (hunk 2 above)
Comment (giuseppecrj) [reply contracts/src/spaces/facets/membership/MembershipFacet.sol:157]: free allocation - value defined by DAO of free memberships you get for a new town
prepaid supply - what a space owner pays for on behalf of their members

File: contracts/src/spaces/facets/prepay/PrepayBase.sol (hunk 3 above)
Comment (giuseppecrj) [reply contracts/src/spaces/facets/prepay/PrepayBase.sol:22]: no, solidity would revert if an underflow happens

**Review by salzbrenner — APPROVED** [review]
I'll do a follow up PR to fix the SpaceDapp. The types aren't broken but the implementation is

**Review by jterzis — APPROVED** [review]

**Pull request merged**